lib/cisco_node_utils/ace.rb

```diff
@@ -0,0 +1,225 @@
+
+module Cisco
+  # Ace - node utility class for Ace Configuration
+  class Ace < NodeUtil
+    attr_reader :acl_name, :seqno, :afi
+    # acl_name is name of acl
+    # seqno is sequence number of ace
+    # afi is either v4 or v6
+    def initialize(acl_name, seqno, afi)
+      fail TypeError unless
+        acl_name.is_a?(String) && seqno.is_a?(Integer) && afi.is_a?(String)
+      fail ArgumentError 'we expect ip or ipv6' unless
+        afi == 'ip' || afi == 'ipv6'
+      @acl_name = acl_name
+      @afi = afi
+      @seqno = seqno
+      @set_args = @get_args = { acl_name: @acl_name, seqno: @seqno, afi: @afi,
+                                src_port: '', dst_port: '',
+                                option_format: '' }
+      @regexp_str = '(?<seqno>\d+) (?<action>\S+)'\
+                 ' *(?<proto>\d+|\S+)'\
+                 ' *(?<src_addr>any|host \S+|\S+\/\d+|\S+ [:\.0-9a-fA-F]+|'\
+                 'addrgroup \S+)*'\
+                 ' *(?<src_port>eq \S+|neq \S+|lt \S+|''gt \S+|range \S+ \S+|'\
+                 'portgroup \S+)?'\
+                 ' *(?<dst_addr>any|host \S+|\S+\/\d+|\S+ [:\.0-9a-fA-F]+|'\
+                 'addrgroup \S+)'\
+                 ' *(?<dst_port>eq \S+|neq \S+|lt \S+|gt \S+|range \S+ \S+|'\
+                 'portgroup \S+)?'\
+                 ' *(?<option>[a-zA-Z0-9\-\/ ]*)*'
+    end
+
+    # Create a hash of all aces under give acl_name.
+    def self.aces
+      regexp_str = '(?<seqno>\d+) (?<action>\S+)'\
+                 ' *(?<proto>\d+|\S+)'\
+                 ' *(?<src_addr>any|host \S+|\S+\/\d+|\S+ [:\.0-9a-fA-F]+|'\
+                 'addrgroup \S+)*'\
+                 ' *(?<src_port>eq \S+|neq \S+|lt \S+|''gt \S+|range \S+ \S+|'\
+                 'portgroup \S+)?'\
+                 ' *(?<dst_addr>any|host \S+|\S+\/\d+|\S+ [:\.0-9a-fA-F]+|'\
+                 'addrgroup \S+)'\
+                 ' *(?<dst_port>eq \S+|neq \S+|lt \S+|gt \S+|range \S+ \S+|'\
+                 'portgroup \S+)?'\
+                 ' *(?<option>[a-zA-Z0-9\-\/ ]*)*'
+      afis = %w(ip ipv6)
+      hash = {}
+      afis.each do |afi|
+        get_args = { afi: afi }
+        instances = config_get('acl', 'all_acl', get_args)
+        next if instances.nil?
+        instances.each do |name|
+          hash[name] = {}
+          get_args[:acl_name] = name
+          aces = config_get('acl', 'all_ace', get_args)
+          next if aces.nil?
+          aces.each do |ace|
+            regexp = Regexp.new(regexp_str)
+            match = regexp.match(ace)
+            item = Ace.new(name, match[1].to_i, afi)
+            item.seqno = match[1].to_i
+            item.action = match[2]
+            item.proto = match[3]
+            item.src_addr = match[4]
+            item.src_port = match[5]
+            item.dst_addr = match[6]
+            item.dst_port = match[7]
+            item.option_format = match[8]
+            hash[name][match[1].to_i] = item
+          end
+        end
+      end
+      hash
+    end
+
```

Comment:
ace_set needs a little more to work right. You can look at bgp_neighbor_af for an example of using set_args_keys helper methods, but this is essentially how it should look:

```
  def ace_set(attrs)
    set_args_keys(attrs)
    @set_args[:state] = attrs.empty? 'no ' : ''
    config_set('acl', 'ace', @set_args, attrs)
  end

def set_args_keys_default
      keys = { asnum: @asn, nbr: @nbr, afi: @afi, safi: @safi }
      keys[:vrf] = @vrf unless @vrf == 'default'
      @get_args = @set_args = keys
    end

    def set_args_keys(hash={})
      set_args_keys_default
      @set_args = @get_args.merge!(hash) unless hash.empty?
    end
```